pinot-broker/src/main/java/org/apache/pinot/broker/routing/adaptiveserverselector/NumInFlightReqSelector.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.adaptiveserverselector;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.pinot.core.transport.server.routing.stats.ServerRoutingStatsManager;
+
+
+/**
+ * The {@code NumInFlightReqSelector} is an AdaptiveServerSelector implementation that picks the best server based on
+ * the number of in-flight queries being processed by the server. The server with the lowest number of in-flight
+ * queries is picked.
+ */
+public class NumInFlightReqSelector implements AdaptiveServerSelector {
+  private final ServerRoutingStatsManager _serverRoutingStatsManager;
+  private final Random _random;
+
+  public NumInFlightReqSelector(ServerRoutingStatsManager serverRoutingStatsManager) {
+    _serverRoutingStatsManager = serverRoutingStatsManager;
+    _random = new Random();
+  }
+
+  @Override
+  public String select(List<String> serverCandidates) {
+    String selectedServer = null;
+    int minValue = Integer.MAX_VALUE;
+
+    // TODO: If two or more servers have same number of in flight requests, break the tie intelligently.
+    for (String server : serverCandidates) {
+      Integer numInFlightRequests = _serverRoutingStatsManager.fetchNumInFlightRequestsForServer(server);
+
+      // No stats for this server. That means this server hasn't received any queries yet.
+      if (numInFlightRequests == null) {
+        int randIdx = _random.nextInt(serverCandidates.size());
+        selectedServer = serverCandidates.get(randIdx);
+        break;
+      }
+
+      if (numInFlightRequests < minValue) {
+        minValue = numInFlightRequests;
+        selectedServer = server;
+      }
+    }
+
+    return selectedServer;
+  }
+
+  @Override
+  public List<String> fetchServerRanking() {
+    List<Pair<String, Double>> pairList = fetchServerRankingWithValues();
```

Comment:
So seems like fetchServerRanking is almost as expensive as fetchServerRankingWithValues + some delta. If the callers of fetchServerRanking can work with name and score/value and retrieve the name from the pair, then we can just unify these two into a single method ?

Reply:
I had initially created 2 for convenience. Dropped `fetchServerRanking()`